rust_icu_ecma402/src/datetimeformat.rs

```diff
@@ -29,13 +29,147 @@ pub struct DateTimeFormat {
+            }
+        }
+        if let Some(ref era) = &opts.era {
+            match era {
+                options::Era::Long => skel.push("GGGG"),
+                options::Era::Short => skel.push("GG"),
+                options::Era::Narrow => skel.push("GGGGG"),
+            }
+        }
+        if let Some(ref year) =&opts.year {
+            match year {
+                options::DisplaySize::Numeric => skel.push("yyyy"),
+                options::DisplaySize::TwoDigit => skel.push("yy"),
+            }
+        }
+        if let Some(ref month) = &opts.month {
+            match month {
+                options::Month::Numeric => skel.push("M"),
+                options::Month::TwoDigit => skel.push("MM"),
+                options::Month::Short => skel.push("MMM"),
+                options::Month::Long => skel.push("MMMM"),
+                options::Month::Narrow => skel.push("MMMMM"),
+            }
+        }
+        if let Some(day) = &opts.day {
+            match day {
+                options::DisplaySize::Numeric => skel.push("d"),
+                options::DisplaySize::TwoDigit => skel.push("dd"),
+            }
+        }
+        // Looks like different formatting is used for 11/12 than 23/24 hours.
+        if let Some(hour) = &opts.hour {
+            if let Some(hour_cycle) = &opts.hour_cycle {
+                match hour_cycle {
+                    options::HourCycle::H11 |
+                    options::HourCycle::H12 => 
+                        match hour {
+                            options::DisplaySize::Numeric => skel.push("h"),
+                            options::DisplaySize::TwoDigit => skel.push("hh"),
+                        }
+                    options::HourCycle::H23 |
+                    options::HourCycle::H24 => match hour {
+                            options::DisplaySize::Numeric => skel.push("H"),
+                            options::DisplaySize::TwoDigit => skel.push("HH"),
+                        }
+                }
+            } else {
+                // This may not be correct for 23/24 hours locale default.
+                match hour {
+                    options::DisplaySize::Numeric => skel.push("h"),
+                    options::DisplaySize::TwoDigit => skel.push("hh"),
+                }
+            }
+        }
+        if let Some(minute) = &opts.minute {
+            match minute {
+                options::DisplaySize::Numeric => skel.push("m"),
+                options::DisplaySize::TwoDigit => skel.push("mm"),
+            }
+        }
+        if let Some(second) = &opts.second {
+            match second {
+                options::DisplaySize::Numeric => skel.push("s"),
+                options::DisplaySize::TwoDigit => skel.push("ss"),
+            }
+        }
+        if let Some(time_zone_style) = &opts.time_zone_style {
+            match time_zone_style {
+                options::TimeZoneStyle::Long => skel.push("VV"),
```

Comment:
Are you sure about time zone styles? According to the syntax guide:

 description | skeleton | output
--- | --- | ---
Time Zone: short time zone ID | V | uslax
Time Zone: long time zone ID | VV | America/Los_Angeles

These are not meant for display to end-users.

Testing `Intl.DateTimeFormat` in the browser, I get

timeZoneName | output | probable skeleton
---|---|---
short | GMT-8 | O
long | Pacific Daylight Time | zzzz

Reply:
Yeah, I'm not 100% sure as to which skeleton needs to be used. :/

Reply:
hm, `uloc_forLanguageTag` fails for any locale that has a country component, but not if the locale has only language.  For example, `sr` works, but `sr-rs` does not.  Weirdness.

Reply:
Yeah, looks like underscores don't work: `en-us` is OK as a language tag but `en_US` isn't.  (Looks like while this has been documented for a while, it only recently started actually being enforced, based on the fact that older ICU builders don't have an issue.)

Reply:
Fixed for timeZoneName.